File: src/Core.CSharp/Consensus.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Zeta.Core.CSharp;

/// <summary>
/// BFT quorum consensus decision core, C# oracle. Conforms to the F# canonical shape
/// (<c>src/Core/Consensus.fs</c>, <c>Consensus.quorumThreshold</c> / <c>decide</c>) by agreeing on the
/// shared seed (<c>src/Core.TypeScript/consensus/golden-vectors.json</c>) that the F#/TS/Rust oracles
/// also verify. Pure integer. The vote state machine is out of byte-lock scope.
/// </summary>
public static class Consensus
{
    /// <summary>The outcome of <see cref="Decide"/>.</summary>
    public sealed record Decision(bool Committed, string? Value, int Count, int Total);

    /// <summary>The classic BFT quorum threshold: <c>2*floor((n-1)/3) + 1</c> (i.e. 2f+1 for n=3f+1).</summary>
    public static int QuorumThreshold(int nodeCount) => (2 * ((nodeCount - 1) / 3)) + 1;

    /// <summary>
    /// Decide consensus over a list of vote values: group by value, take the highest support, commit
    /// it iff it reaches the quorum threshold. The tie-break among values sharing the highest support
    /// is the <b>ordinal minimum</b> — deliberately order-INDEPENDENT, so two nodes that received the
    /// same votes in different orders decide identically. (It was first-occurrence, which read
    /// arrival order and diverged at n in {2,3,6}; see <c>src/Core/Consensus.fs</c> and the shared
    /// seed <c>golden-vectors.json</c>. Do not change it in one oracle.)
    /// </summary>
    public static Decision Decide(IReadOnlyList<string> votes)
    {
        var total = votes.Count;
        if (total == 0)
        {
            return new Decision(false, null, 0, 0);
        }

        var groups = votes
            .GroupBy(v => v, StringComparer.Ordinal)
            .Select(g => (Value: g.Key, Count: g.Count()))
            .ToList();

        var best = groups.Max(x => x.Count);
        // Order-independent tie-break: the ordinal minimum among the values tied at `best`.
        // Ordinal, never culture-sensitive (.claude/rules/culture-invariant-by-default.md).
        var value = groups.Where(x => x.Count == best).Select(x => x.Value).Min(StringComparer.Ordinal)!;

        var threshold = QuorumThreshold(total);
        return best >= threshold
            ? new Decision(true, value, best, total)
            : new Decision(false, null, best, total);
    }
}

File: src/Core.TypeScript/consensus/consensus.ts
// BFT quorum consensus decision core, TypeScript oracle.
// Conforms to the F# canonical shape (src/Core/Consensus.fs, quorumThreshold / decide) by agreeing on
// the shared seed (./golden-vectors.json) that the C#/F#/Rust oracles also verify. Pure integer.
// The vote state machine (transition) carries timestamps and is out of byte-lock scope.

export interface Decision {
  committed: boolean;
  value: string | null;
  count: number;
  total: number;
}

/** The classic BFT quorum threshold: 2*floor((n-1)/3) + 1 (i.e. 2f+1 for n=3f+1). */
export function quorumThreshold(nodeCount: number): number {
  return 2 * Math.trunc((nodeCount - 1) / 3) + 1;
}

/**
 * Decide consensus over a list of vote VALUES: group by value, take the highest support, and commit
 * it iff it reaches quorumThreshold(total).
 *
 * The tie-break among values sharing the highest support is the ORDINAL MINIMUM, deliberately
 * order-INDEPENDENT: two nodes that received the same votes in different orders must decide
 * identically. It used to be first-occurrence, which read arrival order and diverged at
 * n in {2,3,6}. See src/Core/Consensus.fs and the shared seed ./golden-vectors.json — do NOT
 * change the rule in one oracle.
 */
export function decide(votes: string[]): Decision {
  const total = votes.length;
  if (total === 0) return { committed: false, value: null, count: 0, total: 0 };

  const groups: { value: string; count: number }[] = [];
  for (const v of votes) {
    const g = groups.find((x) => x.value === v);
    if (g) g.count++;
    else groups.push({ value: v, count: 1 });
  }

  const best = groups.reduce((m, g) => (g.count > m ? g.count : m), 0);
  // Order-independent tie-break: ordinal minimum among the values tied at `best`. `<` on strings is
  // UTF-16 code-unit (ordinal) comparison — NEVER localeCompare, which is culture-sensitive
  // (.claude/rules/culture-invariant-by-default.md).
  let value = groups.find((g) => g.count === best)!.value;
  for (const g of groups) {
    if (g.count === best && g.value < value) value = g.value;
  }

  const threshold = quorumThreshold(total);
  const top = { value, count: best };
  return top.count >= threshold
    ? { committed: true, value: top.value, count: top.count, total }
    : { committed: false, value: null, count: top.count, total };
}
